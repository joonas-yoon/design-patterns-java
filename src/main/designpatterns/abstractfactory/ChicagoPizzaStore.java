package main.designpatterns.abstractfactory;

import main.designpatterns.abstractfactory.pizza.CheesePizza;
import main.designpatterns.abstractfactory.pizza.PepperoniPizza;
import main.designpatterns.abstractfactory.pizza.VeggiePizza;

import java.util.List;

public class ChicagoPizzaStore extends PizzaStore {

    @Override
    protected Pizza createPizza(String type) {
        FoodFactory factory = new ChicagoFoodFactory();
        return switch (type) {
            case "cheese" -> {
                Pizza pizza = new CheesePizza(factory);
                pizza.setName("Chicago Style Cheese Pizza");
                yield pizza;
            }
            case "pepperoni" -> {
                Pizza pizza = new PepperoniPizza(factory);
                pizza.setName("Chicago Style Pepperoni Pizza");
                yield pizza;
            }
            case "veggies" -> {
                Pizza pizza = new VeggiePizza(factory);
                pizza.setName("Chicago Style Veggie Pizza");
                yield pizza;
            }
            default -> null;
        };
    }
}

class ChicagoFoodFactory implements FoodFactory {
    @Override
    public String createDough() {
        return "Thick deep-dish and buttery dough";
    }

    @Override
    public String createSauce() {
        return "Thin and tangy sauce";
    }

    @Override
    public String createCheese() {
        return "Mozzarella cheese";
    }

    @Override
    public String createVeggies() {
        return "Eggplant, jalapenos";
    }

    @Override
    public List<String> createToppings() {
        return List.of("pepperoni", "sausage", "mushrooms", "onions", "peppers", "olives");
    }
}
