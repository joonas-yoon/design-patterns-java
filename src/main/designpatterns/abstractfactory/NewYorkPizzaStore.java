package main.designpatterns.abstractfactory;

import main.designpatterns.abstractfactory.pizza.CheesePizza;
import main.designpatterns.abstractfactory.pizza.PepperoniPizza;
import main.designpatterns.abstractfactory.pizza.VeggiePizza;

import java.util.List;

public class NewYorkPizzaStore extends PizzaStore {

    @Override
    protected Pizza createPizza(String type) {
        FoodFactory factory = new NewYorkFoodFactory();
        return switch (type) {
            case "cheese" -> {
                Pizza pizza = new CheesePizza(factory);
                pizza.setName("NewYork Style Cheese Pizza");
                yield pizza;
            }
            case "pepperoni" -> {
                Pizza pizza = new PepperoniPizza(factory);
                pizza.setName("NewYork Style Pepperoni Pizza");
                yield pizza;
            }
            case "veggies" -> {
                Pizza pizza = new VeggiePizza(factory);
                pizza.setName("NewYork Style Veggie Pizza");
                yield pizza;
            }
            default -> null;
        };
    }
}

class NewYorkFoodFactory implements FoodFactory {
    @Override
    public String createDough() {
        return "Thin crispy and chewy dough";
    }

    @Override
    public String createSauce() {
        return "Thick and rich sauce";
    }

    @Override
    public String createCheese() {
        return "Mozzarella and parmesan and provolone cheeses";
    }

    @Override
    public String createVeggies() {
        return "Spinach and broccoli and tomatoes";
    }

    @Override
    public List<String> createToppings() {
        return List.of("pepperoni", "sausage", "mushrooms", "onions",
                "peppers", "olives", "green peppers");
    }
}