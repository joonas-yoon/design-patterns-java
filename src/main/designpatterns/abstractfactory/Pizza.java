package main.designpatterns.abstractfactory;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public abstract class Pizza {

    protected String name;
    protected String dough;
    protected String sauce;
    protected String cheese;
    protected String veggies;
    protected List<String> toppings = new ArrayList<>();

    public abstract void prepare();

    public void setName(String name) {
        this.name = name;
    }

    public void cook() {
        if (sauce != null)
            sauce += " (grilled)";
        System.out.println("Cook: " + this);
    }

    public void cut() {
        if (dough != null)
            dough += " (sliced)";
        System.out.println("Cut: " + this);
    }

    @Override
    public String toString() {
        List<String> ingredients = List.of(
                Optional.ofNullable(dough).orElse("No dough"),
                Optional.ofNullable(sauce).orElse("No sauce"),
                Optional.ofNullable(cheese).orElse("No cheese"),
                Optional.ofNullable(veggies).orElse("No veggies")
        );
        return String.format("Pizza {name=%s, ingredients=%s, toppings=%s}",
                name, ingredients, Arrays.toString(toppings.toArray()));
    }
}
