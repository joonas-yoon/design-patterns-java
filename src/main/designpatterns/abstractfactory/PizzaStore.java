package main.designpatterns.abstractfactory;

public abstract class PizzaStore {
    public Pizza orderPizza(String type) {
        Pizza pizza = createPizza(type);
        pizza.prepare();
        pizza.cook();
        pizza.cut();
        return pizza;
    }

    protected abstract Pizza createPizza(String type);
}
