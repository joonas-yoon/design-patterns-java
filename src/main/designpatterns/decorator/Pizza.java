package main.designpatterns.decorator;

import java.util.ArrayList;
import java.util.List;

public abstract class Pizza {

    protected Dough dough = new Dough();
    protected List<String> toppings = new ArrayList<>();

    abstract public int getPrice();

    public String toString() {
        return "{Dough=" + getDough() + ", Toppings=" + getToppings() + "}";
    }

    public List<String> getToppings() {
        return new ArrayList<>(toppings);
    }

    public Dough getDough() {
        return new Dough(dough);
    }
}

