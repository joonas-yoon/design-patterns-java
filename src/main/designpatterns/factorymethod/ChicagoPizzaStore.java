package main.designpatterns.factorymethod;

public class ChicagoPizzaStore extends PizzaStore {

    @Override
    protected Pizza createPizza(String type) {
        return switch (type) {
            case "cheese" -> new ChicagoCheesePizza();
            case "pepperoni" -> new ChicagoPepperoniPizza();
            default -> null;
        };
    }
}

class ChicagoCheesePizza extends ChicagoStylePizza {

    ChicagoCheesePizza() {
        super();
        name = "New York Style Cheese Pizza";
    }
}

class ChicagoPepperoniPizza extends ChicagoStylePizza {

    ChicagoPepperoniPizza() {
        super();
        name = "New York Style Pepperoni Pizza";
        toppings.add("Pepperoni");
    }
}
