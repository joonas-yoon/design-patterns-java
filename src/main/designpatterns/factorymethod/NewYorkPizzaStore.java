package main.designpatterns.factorymethod;

public class NewYorkPizzaStore extends PizzaStore {

    @Override
    protected Pizza createPizza(String type) {
        return switch (type) {
            case "cheese" -> new NewYorkCheesePizza();
            case "pepperoni" -> new NewYorkPepperoniPizza();
            default -> null;
        };
    }
}

class NewYorkCheesePizza extends NewYorkStylePizza {

    NewYorkCheesePizza() {
        super();
        name = "New York Style Cheese Pizza";
    }
}

class NewYorkPepperoniPizza extends NewYorkStylePizza {

    NewYorkPepperoniPizza() {
        super();
        name = "New York Style Pepperoni Pizza";
        toppings.add("Pepperoni");
    }
}
