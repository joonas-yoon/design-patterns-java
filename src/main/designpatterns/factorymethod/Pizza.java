package main.designpatterns.factorymethod;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Pizza {

    String name;
    String dough;
    String sauce;
    List<String> toppings = new ArrayList<>();

    public void cook() {
        sauce += " (grilled)";
        System.out.println("Cook: " + this);
    }

    public void cut() {
        dough += " (sliced)";
        System.out.println("Cut: " + this);
    }

    @Override
    public String toString() {
        return String.format("Pizza {name=%s, dough=%s, sauce=%s, toppings=%s}",
                name, dough, sauce, Arrays.toString(toppings.toArray()));
    }
}
