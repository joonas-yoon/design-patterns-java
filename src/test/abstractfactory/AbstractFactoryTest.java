package test.abstractfactory;

import main.designpatterns.abstractfactory.ChicagoPizzaStore;
import main.designpatterns.abstractfactory.NewYorkPizzaStore;
import main.designpatterns.abstractfactory.Pizza;
import main.designpatterns.abstractfactory.PizzaStore;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import test.BaseUnitTest;

public class AbstractFactoryTest extends BaseUnitTest {

    @Test
    @DisplayName("NewYorkPizzaStore creates pizza")
    public void testOrderPizza() {
        PizzaStore storeAtNewYork = new NewYorkPizzaStore();
        PizzaStore storeAtChicago = new ChicagoPizzaStore();

        println("# Order NY Style cheese pizza");
        Pizza pizzaCheeseNY = storeAtNewYork.orderPizza("cheese");
        println("got pizza: " + pizzaCheeseNY);

        println("# Order Chicago Style cheese pizza");
        Pizza pizzaCheeseCH = storeAtChicago.orderPizza("cheese");
        println("got pizza: " + pizzaCheeseCH);

        println("# Order NY Style pepperoni pizza");
        Pizza pizzaPepperoniNY = storeAtNewYork.orderPizza("pepperoni");
        println("got pizza: " + pizzaPepperoniNY);

        println("# Order Chicago Style pepperoni pizza");
        Pizza pizzaPepperoniCH = storeAtChicago.orderPizza("pepperoni");
        println("got pizza: " + pizzaPepperoniCH);

        println("# Order NY Style veggie pizza");
        Pizza pizzaVeggieNY = storeAtNewYork.orderPizza("veggies");
        println("got pizza: " + pizzaVeggieNY);

        println("# Order Chicago Style veggie pizza");
        Pizza pizzaVeggieCH = storeAtChicago.orderPizza("veggies");
        println("got pizza: " + pizzaVeggieCH);

        assertPrintsTrue();
    }
}
