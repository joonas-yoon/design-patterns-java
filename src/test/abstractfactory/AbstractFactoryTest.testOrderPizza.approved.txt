# Order NY Style cheese pizza
got pizza: Pizza {name=NewYork Style Cheese Pizza, ingredients=[Thin crispy and chewy dough (sliced), Thick and rich sauce (grilled), Mozzarella and parmesan and provolone cheeses, No veggies], toppings=[pepperoni, sausage, mushrooms, onions, peppers, olives, green peppers]}
# Order Chicago Style cheese pizza
got pizza: Pizza {name=Chicago Style Cheese Pizza, ingredients=[Thick deep-dish and buttery dough (sliced), Thin and tangy sauce (grilled), Mozzarella cheese, No veggies], toppings=[pepperoni, sausage, mushrooms, onions, peppers, olives]}
# Order NY Style pepperoni pizza
got pizza: Pizza {name=NewYork Style Pepperoni Pizza, ingredients=[Thin crispy and chewy dough (sliced), Thick and rich sauce (grilled), Mozzarella and parmesan and provolone cheeses, No veggies], toppings=[pepperoni, sausage, mushrooms, onions, peppers, olives, green peppers]}
# Order Chicago Style pepperoni pizza
got pizza: Pizza {name=Chicago Style Pepperoni Pizza, ingredients=[Thick deep-dish and buttery dough (sliced), Thin and tangy sauce (grilled), Mozzarella cheese, No veggies], toppings=[pepperoni, sausage, mushrooms, onions, peppers, olives]}
# Order NY Style veggie pizza
got pizza: Pizza {name=NewYork Style Veggie Pizza, ingredients=[Thin crispy and chewy dough (sliced), Thick and rich sauce (grilled), No cheese, Spinach and broccoli and tomatoes], toppings=[pepperoni, sausage, mushrooms, onions, peppers, olives, green peppers]}
# Order Chicago Style veggie pizza
got pizza: Pizza {name=Chicago Style Veggie Pizza, ingredients=[Thick deep-dish and buttery dough (sliced), Thin and tangy sauce (grilled), No cheese, Eggplant, jalapenos], toppings=[pepperoni, sausage, mushrooms, onions, peppers, olives]}
